File: partition-equal-subset-sum-v4.py
from typing import List


def canPartition(nums: List[int]) -> bool:
    numsSum = sum(nums)

    if len(nums) == 1 or numsSum % 2 == 1:
        return False

    # If both subsets should have the same sum, our target is half the array sum
    targetSum = numsSum // 2
    subsetSums = {0}

    for num in nums:
        moreSubsetSums = set()
        for subsetSum in subsetSums:
            newSum = subsetSum + num
            if newSum == targetSum:
                return True
            moreSubsetSums.add(newSum)
        subsetSums = subsetSums | moreSubsetSums

    return False


nums = [1, 5, 11, 5]
expectedOutput = True
output = canPartition(nums)
assert output == expectedOutput

nums = [1, 2, 3, 5]
expectedOutput = False
output = canPartition(nums)
assert output == expectedOutput

nums = [2, 2, 3, 5]
expectedOutput = False
output = canPartition(nums)
assert output == expectedOutput

File: java/myjava/PartitionEqualSubsetSumV4.java
package myjava;

import java.util.*;

class PartitionEqualSubsetSumV4 {
  public static boolean canPartition(int[] nums) {
    final var numsSum = Arrays.stream(nums).sum();

    if (nums.length == 1 || numsSum % 2 == 1)
      return false;

    // If both subsets should have the same sum, our target is half the array sum
    final var targetSum = numsSum / 2;
    Set<Integer> sumsSet = new HashSet<>();
    sumsSet.add(0);

    for (final var num : nums) {
      Set<Integer> moreSums = new HashSet<>();
      for (final var sum : sumsSet) {
        final var newSum = sum + num;
        if (newSum == targetSum)
          return true;
        moreSums.add(newSum);
      }
      sumsSet.addAll(moreSums);
    }

    return false;
  }

  public static void main(String[] args) {
    assert canPartition(new int[]{1, 5, 11, 5}) == true;
    assert canPartition(new int[]{1, 2, 3, 5}) == false;
    assert canPartition(new int[]{2, 2, 3, 5}) == false;
  }
}
